### ISSUE TITLE
The removeModuleHandlerFactory method has thrown an exception

### ISSUE THREAD

**Issue by openhab-5iver** [eclipse/smarthome#5980]
OH 2.4 snapshot 1319 (ESH 0.10.0.201807241640)

When stopping a manual install of OH (ctrl-D in Karaf), I am seeing this exception. You may need a rule running to reproduce (I haven't tried shutting down without any). I am using the Experimental Rules Engine (JSR223).
```
2018-07-29 11:39:03.364 [DEBUG] [org.eclipse.smarthome.automation.core             ] - ServiceEvent UNREGISTERING - {org.eclipse.smarthome.automation.RuleManager, org.eclipse.smarthome.core.common.registry.RegistryChangeListener}={service.id=525, service.bundleid=211, service.scope=bundle, component.name=org.eclipse.smarthome.automation.core.internal.RuleEngineImpl, component.id=235} - org.eclipse.smarthome.automation.core
2018-07-29 11:39:03.366 [ERROR] [org.eclipse.smarthome.automation.core             ] - [org.eclipse.smarthome.automation.core.internal.RuleEngineImpl(235)] The removeModuleHandlerFactory method has thrown an exception
java.lang.NullPointerException: null
        at org.eclipse.smarthome.automation.core.internal.RuleEngineImpl.getRuleImpl(RuleEngineImpl.java:835) [211:org.eclipse.smarthome.automation.core:0.10.0.201807241640]
        at org.eclipse.smarthome.automation.core.internal.RuleEngineImpl.getStatusInfo(RuleEngineImpl.java:866) [211:org.eclipse.smarthome.automation.core:0.10.0.201807241640]
        at org.eclipse.smarthome.automation.core.internal.RuleEngineImpl.getRuleStatus(RuleEngineImpl.java:1236) [211:org.eclipse.smarthome.automation.core:0.10.0.201807241640]
        at org.eclipse.smarthome.automation.core.internal.RuleEngineImpl.removeMissingModuleTypes(RuleEngineImpl.java:929) [211:org.eclipse.smarthome.automation.core:0.10.0.201807241640]
        at org.eclipse.smarthome.automation.core.internal.RuleEngineImpl.removeModuleHandlerFactory(RuleEngineImpl.java:454) [211:org.eclipse.smarthome.automation.core:0.10.0.201807241640]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
        at org.apache.felix.scr.impl.inject.BaseMethod.invokeMethod(BaseMethod.java:229) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.inject.BaseMethod.access$500(BaseMethod.java:39) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.inject.BaseMethod$Resolved.invoke(BaseMethod.java:650) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.inject.BaseMethod.invoke(BaseMethod.java:506) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.inject.BindMethod.invoke(BindMethod.java:658) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.DependencyManager.invokeUnbindMethod(DependencyManager.java:1837) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.DependencyManager.close(DependencyManager.java:1600) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.SingleComponentManager.disposeImplementationObject(SingleComponentManager.java:353) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.SingleComponentManager.deleteComponent(SingleComponentManager.java:153) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.AbstractComponentManager.doDeactivate(AbstractComponentManager.java:814) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.AbstractComponentManager.deactivateInternal(AbstractComponentManager.java:788) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.DependencyManager.deactivateComponentManager(DependencyManager.java:2225) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.DependencyManager.access$500(DependencyManager.java:55) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.DependencyManager$SingleStaticCustomizer.removedService(DependencyManager.java:1082) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.DependencyManager$SingleStaticCustomizer.removedService(DependencyManager.java:968) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.customizerRemoved(ServiceTracker.java:1241) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.customizerRemoved(ServiceTracker.java:1136) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.ServiceTracker$AbstractTracked.untrack(ServiceTracker.java:996) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:1175) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.BundleComponentActivator$ListenerInfo.serviceChanged(BundleComponentActivator.java:127) [38:org.apache.felix.scr:2.0.12]
        at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:109) [?:?]
        at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:920) [?:?]
        at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230) [?:?]
        at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148) [?:?]
        at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:862) [?:?]
        at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:801) [?:?]
        at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.unregister(ServiceRegistrationImpl.java:222) [?:?]
        at org.apache.felix.scr.impl.manager.AbstractComponentManager$3.unregister(AbstractComponentManager.java:909) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.AbstractComponentManager$3.unregister(AbstractComponentManager.java:874) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.RegistrationManager.changeRegistration(RegistrationManager.java:139) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.AbstractComponentManager.unregisterService(AbstractComponentManager.java:951) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.AbstractComponentManager.doDeactivate(AbstractComponentManager.java:806) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.AbstractComponentManager.deactivateInternal(AbstractComponentManager.java:788) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.AbstractComponentManager.dispose(AbstractComponentManager.java:580) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.manager.ConfigurableComponentHolder.disposeComponents(ConfigurableComponentHolder.java:706) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.BundleComponentActivator.dispose(BundleComponentActivator.java:523) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.Activator.disposeComponents(Activator.java:439) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.Activator.access$300(Activator.java:54) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.scr.impl.Activator$ScrExtension.destroy(Activator.java:293) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.utils.extender.AbstractExtender$2.run(AbstractExtender.java:285) [38:org.apache.felix.scr:2.0.12]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
        at org.apache.felix.utils.extender.AbstractExtender.destroyExtension(AbstractExtender.java:307) [38:org.apache.felix.scr:2.0.12]
        at org.apache.felix.utils.extender.AbstractExtender.bundleChanged(AbstractExtender.java:181) [38:org.apache.felix.scr:2.0.12]
        at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:908) [?:?]
        at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230) [?:?]
        at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148) [?:?]
        at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEventPrivileged(EquinoxEventPublisher.java:213) [?:?]
        at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:120) [?:?]
        at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:112) [?:?]
        at org.eclipse.osgi.internal.framework.EquinoxContainerAdaptor.publishModuleEvent(EquinoxContainerAdaptor.java:168) [?:?]
        at org.eclipse.osgi.container.Module.publishEvent(Module.java:476) [?:?]
        at org.eclipse.osgi.container.Module.doStop(Module.java:634) [?:?]
        at org.eclipse.osgi.container.Module.stop(Module.java:498) [?:?]
        at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.decStartLevel(ModuleContainer.java:1675) [?:?]
        at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1594) [?:?]
        at org.eclipse.osgi.container.SystemModule.stopWorker(SystemModule.java:270) [?:?]
        at org.eclipse.osgi.internal.framework.EquinoxBundle$SystemBundle$EquinoxSystemModule.stopWorker(EquinoxBundle.java:147) [?:?]
        at org.eclipse.osgi.container.Module.doStop(Module.java:636) [?:?]
        at org.eclipse.osgi.container.Module.stop(Module.java:498) [?:?]
        at org.eclipse.osgi.container.SystemModule.stop(SystemModule.java:202) [?:?]
        at org.eclipse.osgi.internal.framework.EquinoxBundle$SystemBundle$EquinoxSystemModule$1.run(EquinoxBundle.java:165) [?:?]
        at java.lang.Thread.run(Thread.java:748) [?:?]
```

**Comment by triller-telekom** [comment]
Since the code where the Exception was thrown was heavily modified in https://github.com/eclipse/smarthome/pull/5942 just one day after the version where the exception occured, I would suggest to leave this issue aside for now and check it again if we have a new ESH stable in openHAB.

Since ESH does not have the karaf console, its not easy to simulate, so please give this a new try after a new ESh stable was build: https://ci.eclipse.org/smarthome/job/SmartHomeDistribution-Stable/

**Comment by kaikreuzer** [comment]
So this exception definitely won't occur this way anymore - I'd therefore suggest to close the issue and open a new one, if a similar issue still exists with the refactored code.

**Comment by openhab-5iver** [comment]
This appears to be resolved in OH snapshot 1325 (ESH 0.10.0.201808011124).
